c/secp256k1_blake160_sighash_all.c

```diff
@@ -28,164 +25,128 @@
   if (ret != CKB_SUCCESS) {
     return ERROR_SYSCALL;
   }
+  if (len > SCRIPT_SIZE) {
+    return ERROR_SCRIPT_TOO_LONG;
+  }
+  mol_pos_t script_pos;
   script_pos.ptr = (const uint8_t*)script;
   script_pos.size = len;
-  args_res = mol_cut(&script_pos, MOL_Script_args());
+
+  mol_read_res_t args_res = mol_cut(&script_pos, MOL_Script_args());
   if (args_res.code != 0) {
     return ERROR_ENCODING;
   }
-  bytes_res = mol_cut_bytes(&args_res.pos);
-  if (bytes_res.code != 0) {
+  mol_read_res_t args_bytes_res = mol_cut_bytes(&args_res.pos);
+  if (args_bytes_res.code != 0) {
     return ERROR_ENCODING;
-  } else if (bytes_res.pos.size == 0) {
-    return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+  } else if (args_bytes_res.pos.size != BLAKE160_SIZE) {
+    return ERROR_ARGUMENTS_LEN;
   }
 
-  /* Calculate sigs count and threshold */
-  sigs_cnt = bytes_res.pos.size / BLAKE160_SIZE;
-  if (bytes_res.pos.size % BLAKE160_SIZE == 0) {
-    threshold = sigs_cnt;
-  } else {
-    threshold = bytes_res.pos.ptr[sigs_cnt * BLAKE160_SIZE];
+  /* Load witness of first input */
+  unsigned char witness[WITNESS_SIZE];
+  len = WITNESS_SIZE;
+  ret = ckb_load_witness(witness, &len, 0, 0, CKB_SOURCE_GROUP_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
   }
-
-  if (threshold > sigs_cnt || threshold == 0) {
-    return ERROR_INVALID_THRESHOLD;
+  if (len > WITNESS_SIZE) {
+    return ERROR_WITNESS_TOO_LONG;
   }
-
-  uint8_t used_signatures[sigs_cnt];
-
-  secp256k1_context context;
-  ret = ckb_secp256k1_custom_verify_only_initialize(&context, secp_data);
-  if (ret != 0) {
-    return ret;
+  if (len < SIGNATURE_SIZE) {
+    return ERROR_WITNESS_TOO_SHORT;
   }
+  size_t extra_witness_len = len - SIGNATURE_SIZE;
 
+  /* Load tx hash */
+  unsigned char tx_hash[BLAKE2B_BLOCK_SIZE];
   len = BLAKE2B_BLOCK_SIZE;
   ret = ckb_load_tx_hash(tx_hash, &len, 0);
   if (ret != CKB_SUCCESS) {
     return ERROR_SYSCALL;
   }
 
+  /* Prepare sign message */
+  unsigned char message[BLAKE2B_BLOCK_SIZE];
+  blake2b_state blake2b_ctx;
+  blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+  blake2b_update(&blake2b_ctx, tx_hash, BLAKE2B_BLOCK_SIZE);
+
+  /* Load extra witness of first input and all witnesses of other inputs */
+  if (extra_witness_len > 0) {
+    blake2b_update(&blake2b_ctx, &witness[SIGNATURE_SIZE], extra_witness_len);
+  }
+  size_t i = 1;
   while (1) {
-    len = 0;
-    /*
-     * Actually we don't need this syscall, we are just making it to grab all
-     * input indices for current group, from which we can load the actual
-     * witness data we need. `since` field is chosen here since it has a fixed
-     * size of 8 bytes, which is both predictable, and also provides minimal
-     * cycle consumption.
-     */
-    ret = ckb_load_input_by_field(NULL, &len, 0, index, CKB_SOURCE_GROUP_INPUT,
-                                  CKB_INPUT_FIELD_SINCE);
+    len = WITNESS_SIZE;
+    ret = ckb_load_witness(temp, &len, 0, i, CKB_SOURCE_INPUT);
```

Comment:
Script should only cover the witness in the group, please use `CKB_SOURCE_GROUP_INPUT`.

Reply:
Updated, and added an unit test, please review again, thanks